source/opt/propagator.cpp

```diff
@@ -0,0 +1,228 @@
+      // Phi arguments come in pairs. Index 'i' contains the
+      // variable id, index 'i + 1' is the originating block id.
+      assert(i % 2 == 0 && i < instr->NumOperands() - 1 &&
+             "malformed Phi arguments");
+
+      uint32_t arg_id = instr->GetSingleWordOperand(i);
+      ir::Instruction* arg_def_instr = get_def_use_mgr()->GetDef(arg_id);
+      uint32_t in_label_id = instr->GetSingleWordOperand(i + 1);
+      ir::Instruction* in_label_instr = get_def_use_mgr()->GetDef(in_label_id);
+      ir::BasicBlock* in_bb = ctx_->get_instr_block(in_label_instr);
+      Edge edge(in_bb, instr_bb);
+
+      if (!IsEdgeExecutable(edge) || ShouldSimulateAgain(arg_def_instr)) {
+        has_operands_to_simulate = true;
+        break;
+      }
+    }
+  } else {
+    instr->ForEachInId([&has_operands_to_simulate, this](const uint32_t* use) {
+      ir::Instruction* def_instr = get_def_use_mgr()->GetDef(*use);
+      if (ShouldSimulateAgain(def_instr)) {
+        has_operands_to_simulate = true;
+        return;
+      }
+    });
+  }
+
+  if (!has_operands_to_simulate) {
+    DontSimulateAgain(instr);
+  }
+
+  return changed;
+}
+
+bool SSAPropagator::Simulate(ir::BasicBlock* block) {
+  if (block == cfg_->pseudo_exit_block()) {
+    return false;
+  }
+
+  // Always simulate Phi instructions, even if we have simulated this block
+  // before. We do this because Phi instructions receive their inputs from
+  // incoming edges. When those edges are marked executable, the corresponding
+  // operand can be simulated.
+  bool changed = false;
+  block->ForEachPhiInst(
+      [&changed, this](ir::Instruction* instr) { changed |= Simulate(instr); });
+
+  // If this is the first time this block is being simulated, simulate every
+  // statement in it.
+  if (!BlockHasBeenSimulated(block)) {
+    block->ForEachInst([this, &changed](ir::Instruction* instr) {
+      if (instr->opcode() != SpvOpPhi) {
+        changed |= Simulate(instr);
+      }
+    });
+
+    MarkBlockSimulated(block);
+
+    // If this block has exactly one successor, mark the edge to its successor
+    // as executable.
+    if (bb_succs_.at(block).size() == 1) {
+      AddControlEdge(bb_succs_.at(block).at(0));
+    }
+  }
+
+  return changed;
+}
+
+void SSAPropagator::Initialize(ir::Function* fn) {
+  // Compute predecessor and successor blocks for every block in |fn|'s CFG.
+  // TODO(dnovillo): Move this to ir::CFG and always build them. Alternately,
+  // move it to IRContext and build CFG preds/succs on-demand.
+  bb_succs_[cfg_->pseudo_entry_block()].push_back(
+      Edge(cfg_->pseudo_entry_block(), fn->entry().get()));
+
+  for (auto& block : *fn) {
+    block.ForEachSuccessorLabel([this, &block](uint32_t label_id) {
+      ir::BasicBlock* succ_bb =
+          ctx_->get_instr_block(get_def_use_mgr()->GetDef(label_id));
+      bb_succs_[&block].push_back(Edge(&block, succ_bb));
+      bb_preds_[succ_bb].push_back(Edge(succ_bb, &block));
```

Comment:
I don't see bb_preds_ being used anywhere. Should bb_preds_ also feature pseudo-blocks?

Reply:
Good point.  Added.  Predecessors are not really used in the propagator, but I'm thinking of moving this into ir::CFG and I *think* I may need it in SSA-CCP itself.